### PR TITLE
Download feature for single episode

### PR DESCRIPTION
**Pull request by ypawar159**
Hey @DxCx ,
I have rewritten the code considering the feedback of previous PR #187 
Let me know if anything else required.
Will work on feature to add more flexible download options once this is merged.

### CODE REVIEW

File: default.py [hunk 1]
```diff
@@ -146,18 +167,36 @@ def PLAY(payload):
         if s._read_sources():
             items = sorted(s._sources.iteritems(), key=lambda x: x[0])
             items = [(title[5:], url) for title, url in items]
-            items = map(lambda x: utils.allocate_item(x[0], 'playlink&url=/'+x[1], False, ''), items)
+            # items = map(lambda x: utils.allocate_item(x[0], 'playlink&url=/'+x[1], False, ''), items)
```
Comment (DxCx) [inline default.py]: cleanup =)

File: resources/lib/ui/control.py [hunk 2]
```diff
@@ -104,3 +120,37 @@ def draw_items(video_data):
             xbmc_add_player_item(vid['name'], vid['url'], vid['image'])
     xbmcplugin.endOfDirectory(HANDLE, succeeded=True, updateListing=False, cacheToDisc=True)
     return True
+
+
+def log(msg, level="INFO"):
```
Comment (DxCx) [inline resources/lib/ui/control.py]: please remove this from here
we will tackle this in the other PR (#205 )

File: resources/lib/ui/embed_extractor.py [hunk 3]
```diff
@@ -97,7 +97,7 @@ def playSource():
         reqObj = http.send_request(playUrl)
         return VIDEO_RE.findall(reqObj.text)[0]
 
-    return playSource
+    return playSource()
```
Comment (DxCx) [inline resources/lib/ui/embed_extractor.py]: adn this ...

File: resources/lib/ui/control.py [hunk 4]
```diff
@@ -104,3 +120,37 @@ def draw_items(video_data):
...
+    if level == "ERROR":
+        xbmc.log(msg=log_message, level=xbmc.LOGERROR)
+    if level == "SEVERE":
+        xbmc.log(msg=log_message, level=xbmc.LOGSEVERE)
+    if level == "FATAL":
+        xbmc.log(msg=log_message, level=xbmc.LOGFATAL)
+
+
+def aria2_download(url, fname, path):
+    aria_url = "http://{}:{}/rpc".format(getSetting('ipaddress'), getSetting('port'))
```
Comment (DxCx) [inline resources/lib/ui/control.py]: to solve the playSource() thingy you want to add here:
linkInfo = _prefetch_play_link(url)

then use linkInfo['url']

File: default.py [hunk 5]
```diff
@@ -146,18 +167,36 @@ def PLAY(payload):
...
             return control.draw_items(items)
     else:
         return control.play_source(s.get_video_link())
 
-@route('playlink*')
+
+@route('playlink/*')
 def PLAY_SOURCE(payload):
-    return control.play_source(urlparse.unquote(payload))
+    par = payload.split("&")
```
Comment (DxCx) [inline default.py]: lets use urlparse for this:
```python
>>> from urlparse import urlparse, parse_qs
>>> URL='https://someurl.com/with/query_string?i=main&mode=front&sid=12ab&enc=+Hello'
>>> parsed_url = urlparse(URL)
>>> parse_qs(parsed_url.query)
{'i': ['main'], 'enc': [' Hello '], 'mode': ['front'], 'sid': ['12ab']}
```

less room for errors :)

**Comment by ypawar159** [comment]
@DxCx ,
Will do after this weekend.

**Pull request closed without merging**